044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,345 @@
+plugin instead and block metadata-generating operations.
+
+*tested with kafka 3.3.1
+### High level changes
+
+To better support alternative sources for managing quotas and how they interact with volume usage this proposal
+separates some existing concepts and separates their responsibilities. We envisage the architecture looking like
+this ![Component interactions](images/044-quota-plugin-interactions.png)
+
+#### Quota Source
+
+We propose adding a Quota Source concept to the plugin to provide an abstraction through which where we could provide
+external sources of quotas, like pull them from an external system.
+
+#### Throttle Factor
+
+We introduce the concept of throttle factor. This is a factor in the range (0.0, 1.0) that is applied to a quota to
+calculate the final byte rate limit. 
+
+This is a concept that was already implicit in the current calculations, but we want to name it so that we can use it in 
+metric names and provide an extension point in case we want to externalise the Throttle Factor.
+
+#### Throttle Factor Source
+
+We propose adding a Throttle Factor Source concept to the plugin to provide an abstraction through which we could
+provide alternative sources of throttle factors, such as pull the factor from a single decision-making broker.
+
+#### Volume
+
+The storage quotas operate on observations about a **Volume** with these characteristics:
+1. `logDir`: the path of the logdir
+2. `brokerId`
+3. `totalBytes`
+4. `availableBytes`
+5. `usedBytes` (totalBytes - availableBytes)
+6. `availableRatio` (availableBytes/totalBytes)
+
+#### Volume Source
+
+The quota plugin operates on observations about Volumes, the **Volume Source** determines where
+we obtain those observations. Values are:
+1. `local`: we inspect this brokers local log dirs to discover the state of their Volume
+2. `cluster`: we ask kafka for the state of all log dirs for all currently active nodes
+
+the source will be configurable with a property like:
+
+1. `client.quota.callback.static.storage.volume.source=cluster`
+
+Local will be the default if the property isn't provided but is deprecated from the beginning.
+
+#### Cluster Volume Source
+
+With the introduction of [KIP-827](https://cwiki.apache.org/confluence/display/KAFKA/KIP-827%3A+Expose+log+dirs+total+and+usable+space+via+Kafka+API) in
+kafka 3.3 we can now obtain the total and usable (available) bytes per log dir as part of the DescribeLogDirsResponse.
+
+Note: if a single disk contains multiple log dirs, it will be described multiple times through the kafka APIs. This
+repetition is acceptable as our new limit types will be applied per-volume, so redundant volume descriptions don't
+impact the outcome.
+
+The Cluster Volume Source will use this API to discover volume information for the whole cluster. We intend to continue
+using `client.quota.callback.static.storage.check-interval` to configure the millisecond interval between polling the
+cluster state.
```

Comment:
nit: Since we've mentioned `client.quota.callback.static.storage.check-interval` in the following paragraph, I think this paragraph is a little redundant. WDYT?

Reply:
Merged the two paragraphs to avoid the redundancy.